python/tests/test_umap.py

```diff
@@ -0,0 +1,226 @@
+
+    else:  # dataset == "iris"
+        local_X, local_y = load_iris(return_X_y=True)
+
+    local_X = cp.asarray(local_X)
+    local_y = cp.asarray(local_y)
+
+    local_X = local_X.repeat(math.ceil(n_rows / len(local_X)), axis=0)
+    local_y = local_y.repeat(math.ceil(n_rows / len(local_y)), axis=0)
+
+    # Add some gaussian noise
+    local_X += cp.random.standard_normal(local_X.shape, dtype=cp.float32)
+
+    return local_X, local_y
+
+
+def _local_umap_trustworthiness(
+    local_X: np.ndarray,
+    local_y: np.ndarray,
+    n_neighbors: int,
+    supervised: bool,
+) -> float:
+    from cuml.manifold import UMAP
+
+    local_model = UMAP(n_neighbors=n_neighbors, random_state=42, init="random")
+    y_train = local_y if supervised else None
+    local_model.fit(local_X, y=y_train)
+
+    embedding = local_model.embedding_
+    return trustworthiness(local_X, embedding, n_neighbors=n_neighbors, batch_size=5000)
+
+
+def _spark_umap_trustworthiness(
+    local_X: np.ndarray,
+    local_y: np.ndarray,
+    n_neighbors: int,
+    supervised: bool,
+    n_parts: int,
+    n_workers: int,
+    sampling_ratio: float,
+    dtype: np.dtype,
+    feature_type: str,
+) -> float:
+    local_model = UMAP(
+        n_neighbors=n_neighbors,
+        sample_fraction=sampling_ratio,
+        random_state=42,
+        init="random",
+        num_workers=n_workers,
+    )
+
+    with CleanSparkSession() as spark:
+        if supervised:
+            data_df, features_col, label_col = create_pyspark_dataframe(
+                spark, feature_type, dtype, local_X, local_y
+            )
+            assert label_col is not None
+            local_model.setLabelCol(label_col)
+        else:
+            data_df, features_col, _ = create_pyspark_dataframe(
+                spark, feature_type, dtype, local_X, None
+            )
+
+        data_df = data_df.repartition(n_parts)
+        local_model.setFeaturesCol(features_col)
+        distributed_model = local_model.fit(data_df)
+        embedding = cp.array(distributed_model.embedding_)
+
+    return trustworthiness(local_X, embedding, n_neighbors=n_neighbors, batch_size=5000)
+
+
+def _run_spark_test(
+    n_parts: int,
+    n_workers: int,
+    n_rows: int,
+    sampling_ratio: float,
+    supervised: bool,
+    dataset: str,
+    n_neighbors: int,
+    dtype: np.dtype,
+    feature_type: str,
+) -> bool:
+    local_X, local_y = _load_dataset(dataset, n_rows)
+
+    dist_umap = _spark_umap_trustworthiness(
+        local_X,
+        local_y,
+        n_neighbors,
+        supervised,
+        n_parts,
+        n_workers,
+        sampling_ratio,
+        dtype,
+        feature_type,
+    )
+
+    loc_umap = _local_umap_trustworthiness(local_X, local_y, n_neighbors, supervised)
+
+    print("Local UMAP trustworthiness score : {:.2f}".format(loc_umap))
+    print("Spark UMAP trustworthiness score : {:.2f}".format(dist_umap))
+
+    trust_diff = loc_umap - dist_umap
+
+    return trust_diff <= 0.15
```

Comment:
Is there any intuition to set the tolerance to 0.15 and not other value?  Will be good to add a comment to explain.

Reply:
This is taken from the cuml unit tests, so hopefully it is based on some empirical testing on their part. The trustworthiness metric ranges from 0-1, and given the stochasticity of transform, some lenience makes sense. Not sure if 0.15 is the ideal value though, they don't have any written justification.